File: getting_started/insertion_sort/insertion_sort_python/insertion_sort.py
"""
- Implementation of the insertion sort algorithm in python
- Author: Abdoulaye Diallo
- Complexity O(n^2)
"""


def insertionSort(a):
    """ Implementation in python of the insertion sort algorithm 
        - Complexity O(n²)
    """
    for j in range(1, len(a)):
        key = a[j]
        i = j - 1
        while (i > -1 and a[i] > key):
            a[i+1] = a[i]
            i = i-1
        a[i+1] = key

    return a 

if __name__ == '__main__':
    arr = [5, 2, 4, 6, 1, 3]
    print(insertionSort(arr))

File: getting_started/insertion_sort/insertion_sort_c/insertion_sort.c
/**
 * Implementation of the insertion sort algorithm in C
 * Author : Abdoulaye Diallo
 * Complexity O(n^2)
 **/

#include <stdio.h>
#include <stdlib.h>

void insertionSort(int *arr, int n);
void printArray(int *arr, int n);

int main()
{
    int arr[] = {5, 3, 4, 6, 2, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    insertionSort(arr, n);

    printArray(arr, n);
    return EXIT_SUCCESS;
}

void insertionSort(int *arr, int n) {
    int i, j, key;
    for (j = 1; j < n; j++) {
	key = arr[j];
	i = j-1;

	while (i > -1 && arr[i] > key) {
	    arr[i+1] = arr[i];
	    i = i-1;
	}

	arr[i+1] = key;
    }
}

void printArray(int *arr, int n) {
    int i;
    for (i = 0; i < n; i++)
	printf("%d ", arr[i]);
    
    printf("\n");
}
